### PR TITLE
feat: --mergeConfig option for --deep to inherit config settings

### PR DESCRIPTION
**Pull request by art-ws**
#895

### CODE REVIEW

File: lib/index.js [hunk 1]
```diff
@@ -508,9 +508,14 @@ async function run(options = {}) {
         const packages = await previousPromise
         // copy object to prevent share .ncurc options between different packageFile, to prevent unpredictable behavior
         const rcResult = getNcurc({ packageFile })
+        const rcConfig = rcResult && rcResult.config ? rcResult.config : {}
+        if (options.recursive && options.reject) {
+          // inherit parent config options
+          rcConfig.reject = [...new Set([...options.reject, ...rcConfig.reject || []])]
```
Comment (raineorshine) [inline lib/index.js:514]: We should apply config inheritance to all options, not just reject.

**Review by raineorshine — CHANGES_REQUESTED** [review]

File: lib/index.js (hunk 1 above)
Comment (art-ws) [reply lib/index.js:514]: Yep. Can you please advise where to find the complete list of nested structured properties?

Comment (raineorshine) [reply lib/index.js:514]: All of the options are listed in [https://github.com/raineorshine/npm-check-updates/blob/main/lib/cli-options.js](cli-options.js). I hope you’re not planning on applying them individually?

Comment (art-ws) [reply lib/index.js:514]: For merge we can apply such simple generic function with test coverage:
```javascript
function mergeArrays(arr1, arr2){
  return [...new Set([...(arr1 || []), ...(arr2 || [])])]
}

function mergeOptions(options1, options2){
  options1 = options1 || {}
  options2 = options2 || {}  
  const result = {...options1, ...options2 }
  for (const key in result) {
    if (result.hasOwnProperty(key)) {
      if (Array.isArray(options1[key]) && Array.isArray(options2[key])){
        result[key] = mergeArrays(options1[key], options2[key]) 
      }
    }
  }
  return result;
}
```

i.e. it is a normal merge, but if both properties are arrays, then we merge them also.
WDT?

Comment (raineorshine) [reply lib/index.js]: Sure, I'm okay with that.

File: lib/index.js [hunk 2]
```diff
@@ -587,4 +611,4 @@ function getNcurc({ configFileName, configFilePath, packageFile } = {}) {
   return result ? { ...result, args } : null
 }
 
-module.exports = { run, getNcurc, getPeerDependencies, ...vm }
+module.exports = { run, getNcurc, getPeerDependencies, mergeOptions, ...vm }
```
Comment (raineorshine) [inline lib/index.js:614]: We don't want to add `mergeOptions` to the public api just for testing. Instead, define it in its own module in `/src/lib` and import it directly in the test.

File: lib/index.js [hunk 3]
```diff
@@ -46,6 +46,25 @@ function programError(options, message) {
   }
 }
 
+function mergeArrays(arr1, arr2) {
+  return [...new Set([...arr1 || [], ...arr2 || []])]
+}
+
+/**
+ * Merge options by a standard way, but if some properties both are arrays, then merge them also.
```
Comment (raineorshine) [inline lib/index.js:54]: You should specify in the JSDOC that is a shallow merge (as opposed to a recursive or deep merge).

File: test/deep.test.js [hunk 4]
```diff
@@ -144,6 +144,67 @@ describe('--deep with nested ncurc files', function () {
...
+    deepJsonOut.should.have.property('pkg/sub3/package.json')
+    deepJsonOut['pkg/sub3/package.json'].should.not.have.property('cute-animals')
+    deepJsonOut['pkg/sub3/package.json'].should.have.property('fp-and-or')
+    deepJsonOut['pkg/sub3/package.json'].should.have.property('ncu-test-v2')
+  })
+
+  it('use ncurc of nested packages with --recursive option', async () => {
+
+    const configFilePath = path.join(__dirname, 'deep-ncurc', '.ncurc.js')
+    const deepJsonOut = await spawn('node', [bin, '--jsonUpgraded', '--deep', '--recursive', '--configFilePath', configFilePath], { cwd }).then(JSON.parse)
```
Comment (raineorshine) [inline test/deep.test.js:158]: If you specify `--configFilePath` on the command line, it will use that config file on every nested package anyway (without `--recursive`), as command line options apply to every package. If you omit this option, it will use the config file in `cwd`.

File: lib/index.js [hunk 5]
```diff
@@ -508,9 +527,14 @@ async function run(options = {}) {
         const packages = await previousPromise
         // copy object to prevent share .ncurc options between different packageFile, to prevent unpredictable behavior
         const rcResult = getNcurc({ packageFile })
+        let rcConfig = rcResult && rcResult.config ? rcResult.config : {}
+        if (options.recursive) {
+          // inherit parent config options
+          rcConfig = mergeOptions(options, rcConfig)
```
Comment (raineorshine) [inline lib/index.js:533]: This inherits not just the parent's config options, but the parent's command-line options. We currently maintain this distinction:

- command line options apply to every package
- config options apply locally to the package they are in

Merging command-line options instead of just merging config options might have unintended consequences.

**Review by raineorshine — CHANGES_REQUESTED** [review]

**Comment by raineorshine** [comment]
I suspect that there may be another problem with this logic and that it is not exposed by the test case. I could be wrong but I wanted to bring it up in case you're not aware of it.

Subpackages are retrieved as a flat array of paths via `globby.sync`:

https://github.com/raineorshine/npm-check-updates/blob/486535fd605ceb8cf402029dde784bb404e36c0c/lib/index.js#L492-L497

This means that there is no notion of "parent" and "child" in the current `--deep` implementation. Practically this means that with a structure such as the following:

- a
  - b
    - c
  - d
    - e

There is no way in the current implementation to tell that "e" should inherit from "a" and "d" but not "b" or "c". I'm guessing that it will fail on such a test case.

You may have to replace globby with manual traversal in order to guarantee the order and know who the parent is.

File: lib/index.js (hunk 2 above)
Comment (art-ws) [reply lib/index.js:614]: OK

File: test/deep.test.js (hunk 4 above)
Comment (art-ws) [reply test/deep.test.js:158]: Without `--configFilePath` tests not works as implementation have a side-effects depended on `process.env.NCU_TESTS`
and internal state is diffrent due:
```javascript
// load .ncurc
// Do not load when global option is set
// Do not load when tests are running (an be overridden if configFilePath is set explicitly)
const rcResult = !program.global && (!process.env.NCU_TESTS || configFilePath)
  ? ncu.getNcurc({ configFileName, configFilePath, packageFile })
  : null
```
Behevior is very implicit and not obvious. Same code rinning from cli and from test has a different behevior.
As a lot of tests depends on `process.env.NCU_TESTS` it was a most simple way to arcjive the predictable behevior.

File: lib/index.js (hunk 3 above)
Comment (art-ws) [reply lib/index.js:54]: OK

File: lib/index.js (hunk 5 above)
Comment (art-ws) [reply lib/index.js:533]: Yes, you right. Will change implementation to merge onlu subset of keys from `rcConfig`

**Comment by art-ws** [comment]
> I suspect that there may be another problem with this logic and that it is not exposed by the test case. I could be wrong but I wanted to bring it up in case you're not aware of it.

"Perfect is the enemy of the good" :) Actually, such enhancement  already successfully cover the practical cases and generally bring a new step in overall tool evolution. But, for sure, make sense to think how to cover more complex cases.

File: test/deep.test.js (hunk 4 above)
Comment (raineorshine) [reply test/deep.test.js]: > Behevior is very implicit and not obvious. Same code rinning from cli and from test has a different behevior.
> As a lot of tests depends on `process.env.NCU_TESTS` it was a most simple way to achieve the predictable behevior without core code modifications.

I agree, not the best solution. The tests should probably always run in a clean directory.

I like your solution of checking `recursive` to ensure the config file gets loaded.

File: lib/index.js (hunk 5 above)
Comment (raineorshine) [reply lib/index.js]: That's not quite what I had in mind. I will try to explain in a review comment!

File: lib/index.js [hunk 6]
```diff
@@ -508,9 +509,14 @@ async function run(options = {}) {
         const packages = await previousPromise
         // copy object to prevent share .ncurc options between different packageFile, to prevent unpredictable behavior
         const rcResult = getNcurc({ packageFile })
+        let rcConfig = rcResult && rcResult.config ? rcResult.config : {}
+        if (options.recursive) {
+          // Merge config options
+          rcConfig = mergeOptions(options, rcConfig, { keys: Object.keys(rcConfig) })
```
Comment (raineorshine) [inline lib/index.js:515]: `options` consist of command-line options and config options which are combined (prematurely imo) in `cli.js`. Thus we don't have direct access to the root package config at this point. My previous comment was saying that ideally we only merge config options (not command line options) but that doesn't seem possible without some additional refactoring. Luckily it doesn't seem to be a problem with the main use case we are solving here.

So, I may not have been following the `key` selection logic correctly, but I don't think we need it now and I don't think it offers much advantage if we are merging `options` anyway. Let me know if I'm missing something!

**Review by raineorshine — CHANGES_REQUESTED** [review]

**Comment by raineorshine** [comment]
Considering the functionality that we've settled on, I think a more appropriate name would be `--merge-config`. `--deep` is already recursive and this option specifically merges nested configs with the root config rather than overriding. What do you think?

Make sure to run `npm run build` to regenerate the README and type definitions.

File: lib/index.js (hunk 6 above)
Comment (art-ws) [reply lib/index.js:515]: Technically `{ keys: Object.keys(rcConf) }` guarantee(we have test coverage on it) that we merge only the properties from  `rcResult.config` and all other properties from `options` will not be changed. For our particular use case it may be even more explicit `{ keys: ['reject'] }`, but current implementation is more generic and looks pretty safe. Logic behind `mergeOptions` is simple - it's the same `Object.assign` but if some props are arrays, then we merge them also. Additional  `keys` parameter allow us to be sure that may be changed only properties explicitly specified in `rcConfig` to prevent  unintended consequences.

Any case all this feature isolated from all other code with `--deep` and `--recursive` options, so this enhancement fix the problem and not brings the breaking changes.

Comment (raineorshine) [reply lib/index.js:515]: > Technically `{ keys: Object.keys(rcConf) }` guarantee(we have test coverage on it) that we merge only the properties from `rcResult.config` and all other properties from `options` will not be changed.

I'm not sure I understand. If an option does not exist in `rcResult`, then how could it overwrite `options`?

i.e. `_.pick(options2, Object.keys(options2))` is a noop

Comment (art-ws) [reply lib/index.js:515]: Correct. If no any properties in `rcResult` then nothing to merge for this config, i.e. no need to override something in `options`. The idea is to override only properties, which explicitly defined in `rcResult`. If nothing defined, then nothing to do.

Comment (raineorshine) [reply lib/index.js:515]: Thanks for your response. I was unclear in my last message, so let me try again. Please correct me if I am missing something.

What I'm seeing is that on L515 you call:


```js
mergeOptions(options, rcConfig, { keys: Object.keys(rcConfig) })
```

And in merge-options [L15](https://github.com/raineorshine/npm-check-updates/pull/897/files#diff-c37a68c010b04a2f15f188f945826efc4a558a607f7d1871fda1d557ce4d25ffR15) we have: 

```js
options2 = opts.keys ? _.pick(options2 || {}, opts.keys) : options2 || {} 
```

So when the arguments get substituted through, we get:

```js
options2 = opts.keys ? _.pick(rcConfig, Object.keys(rcConfig) : rcConfig
```

Which is always `options2 = options2`, regardless of the inputs. Thus the `keys` logic is entirely superfluous which I don't believe was your intention. If you could remove `opts.keys` I think everything else looks good to go.

**Review by raineorshine — CHANGES_REQUESTED** [review]

**Comment by art-ws** [comment]
Oh..yes:) Good catch. Such things happens when 2 diffrent ideas meets together during the series of refactoring. 
The purpose of `keys` argument was to have a generic approach to restrict keys to merge, but in case of actual usage of such function we always have a trivial case.

Let's simplify implementation.

**Comment by raineorshine** [comment]
Thanks! I’ll take a look asap

**Review by raineorshine — APPROVED** [review]

**Comment by raineorshine** [comment]
Published in `v11.6.0`
